### PR TITLE
[Logs]: Kubernetes integration

### PR DESCRIPTION
**Pull request by mnshdw**
### What does this PR do?

This PR adds an experimental log collection feature for kubernetes environments that relies on either kubelet or inotify to detect pods, rather than the Docker api.

This feature is currently enabled by setting the environment variable `DD_LOGS_CONFIG_CONTAINER_COLLECT_ALL` to `true`. <= This has to change before making the PR final.

### Motivation

Collect container logs in Kubernetes environments that do not use Docker as a container runtime.

### CODE REVIEW

**Review by ajacquemot — APPROVED** [review]

**Review by ajacquemot — DISMISSED** [review]

File: pkg/logs/config/config_test.go [hunk 1]
```diff
@@ -21,36 +21,49 @@ func TestDefaultDatadogConfig(t *testing.T) {
...
 	assert.Equal(t, 100, LogsAgent.GetInt("logs_config.open_files_limit"))
+	assert.Equal(t, false, LogsAgent.GetBool("logs_config.container_collect_all"))
 }
 
 func TestBuildLogsSources(t *testing.T) {
 	var ddconfdPath string
 	var logsSources *LogSources
-	var source *LogSource
+
+	// default tail all containers source should be the last element of the list
```
Comment (tmichelet) [inline pkg/logs/config/config_test.go]: should we test this?

File: pkg/logs/input/container/scanner.go [hunk 2]
```diff
@@ -3,207 +3,30 @@
...
-		if !isTailed {
-			// setup a new tailer
-			succeeded := s.setupTailer(s.cli, container, source, tailFromBeginning, s.pp.NextPipelineChan())
-			if !succeeded {
-				// the setup failed, let's try to tail this container in the next scan
-				continue
-			}
-		}
-		containersToMonitor[container.ID] = true
+		sources.AddSource(config.NewLogSource("container_collect_all", dockerConfig))
```
Comment (tmichelet) [inline pkg/logs/input/container/scanner.go]: we're altering the original list of sources, could that be a problem in the future?

File: pkg/logs/input/container/kubernetes_scanner.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+const kubernetesIntegration = "kubernetes"
+
+// getSource returns a new source for the container in pod
+func (s *KubeScanner) getSource(pod *kubelet.Pod, container kubelet.ContainerStatus) *config.LogSource {
+	return config.NewLogSource(s.getSourceName(pod, container), &config.LogsConfig{
+		Type:    config.FileType,
+		Path:    s.getPath(pod, container),
+		Source:  kubernetesIntegration,
+		Service: kubernetesIntegration,
+		Tags:    s.getTags(container),
```
Comment (tmichelet) [inline pkg/logs/input/container/kubernetes_scanner.go]: Are tags updates possible? Would this catch any tags updates?

File: pkg/logs/input/container/kubernetes_scanner.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+// addSources creates a new log source for each container of a new pod.
+func (s *KubeScanner) addSources(pod *kubelet.Pod) {
+	for _, container := range pod.Status.Containers {
+		containerID := container.ID
+		if _, exists := s.sourcesByContainer[containerID]; exists {
+			continue
+		}
+		source := s.getSource(pod, container)
+		s.sourcesByContainer[containerID] = source
+		s.sources.AddSource(source)
```
Comment (tmichelet) [inline pkg/logs/input/container/kubernetes_scanner.go]: don't we need to start tailing the logs?

Comment (tmichelet) [reply pkg/logs/input/container/kubernetes_scanner.go]: Oh, I get it, we add a file source to the list of sources so the file scanner picks up the tailing. Definitely worth a comment here, and an explanation at the beginning of the class

**Review by tmichelet — COMMENTED** [review]
Nice cleanup, I wish we did it in two steps. From what I see, this PR:
- does a bunch of cleaning 
- updates the behavior of collection of docker logs in k8 environment which is riskier and could fit a second PR to be easily reviewed and reverted / patched if needed.

That said, about the k8 udpate:
- I think we should have a config flag to pick between kubelet and docker api, so users can switch from one to the other if they need to revert for any reason
- I don't like the idea of mixing sources that come from configuration with sources that come from discovery. I might be ok with mixing them in the Status, but even there I think in the future we'll want different ways of visualizing the info depending on if it's discovered or set up. But for the different scanners, I don't feel good about manipulating those together. Maybe we could extract the file tailing / file scanning logic, and then have two file scanners, one for File configs and one for kubernetes? Thoughts?

**Pull request closed without merging**